### PR TITLE
Testing setup

### PR DESCRIPTION
**Pull request by janvandermeulen**
In this PR testing for the model-training is added. 
- Unit tests are added for the first parts of the pipeline is added to ensure the functions work properly. 
    - These tests are done on a slice of the real dataset for speedup. 
- Integration tests are added to test the functionality of the full learning pipeline.
- Tests for data uniqueness are added to ensure the dataset is of high enough quality.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -2,9 +2,11 @@
 name = "phishing-detection"
 version = "0.1.0"
 description = ""
-
 authors = ["Jan <j2000.vdm@gmail.com>", "Shayan Ramezani <s.ramezani@student.tudelft.nl>", "Remi Lejeune <R.J.Lejeune@student.tudelft.nl", "Michael Chan <J.M.Chan@student.tudelft.nl>"]
 readme = "README.md"
+packages = [{ include = "src" },
+            { include = "tests" }
```
Comment (sr1998) [inline pyproject.toml]: I don't think `tests` should be in packaging logic for poetry

File: src/run.py [hunk 2]
```diff
@@ -36,32 +37,39 @@ def run(params: dict, data_path: str) -> None:
         params = yaml.safe_load(path)
         run(params)
     """
-    train = load_data_from_text(os.path.join(data_path, "train.txt"))
-    test = load_data_from_text(os.path.join(data_path, "test.txt"))
-    val = load_data_from_text(os.path.join(data_path, "val.txt"))
+    path = os.path.join(os.path.dirname(__file__), "../", data_path)
```
Comment (sr1998) [inline src/run.py:40]: `../` should be `..` right? os.path.join takes care of adding `/`

Maybe there is even logic in python for `..`

File: tests/test_preprocess.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+from src import preprocess
+import numpy as np
+import pytest
+import json
+
+
+@pytest.mark.fast
+def test_fake_data():
+    sys.argv = ["", "tests/testdata"]
+    preprocess.main()
```
Comment (sr1998) [inline tests/test_preprocess.py:14]: Open for discussion: But isn't it better to have fixtures or some other logic for saving/loading or even hardcoding data?

Some online discussions: 
https://stackoverflow.com/questions/34466027/what-is-conftest-py-for-in-pytest
https://stackoverflow.com/questions/54330834/what-is-a-correct-approach-to-manage-test-data-using-pytest

Repeating paths to test data is a pain. We can at least have those hardcoded for pytest in ways that pytest makes possible for us.

File: tests/test_integration.py [hunk 4]
```diff
@@ -1,6 +1,63 @@
...
+# Set the path to the src directory
+sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), '../src')))
+from src import model_definition
+from src import preprocess
+from src import predict
+from src import train
+import numpy as np
 import pytest
+import yaml
+import json
```
Comment (sr1998) [inline tests/test_integration.py:14]: About the markers. The pipeline is hardcoded on the markers (which I guess we could have a file to keep track of them all or maybe pytest provides some logic for). But anyhow, I kinda liked unit and integration, but also like fast and slow. The question is, is slow meant only for integration or also for other kind of tests we may add later? What defines speed? 

Previously, the integration tests were only ran when PR is open with main. Unittest are ran for any push and PR. The question is, do we want the same thing for fast and slow, or will we have some slow ones on every push? 

In general, I think having specific markers is better, if it doesn't get out of hand.

**Review by sr1998 — COMMENTED** [review]
Looks good. Left some comment

File: tests/test_integration.py (hunk 4 above)
Comment (MichaelChan20) [reply tests/test_integration.py:14]: I think we could also have markers based on the 4 angles described in the lecture.

File: src/run.py (hunk 2 above)
Comment (janvandermeulen) [reply src/run.py:40]: ```../``` Describes the logic more intuitively in my opinion, even though both are correct of course.

File: pyproject.toml (hunk 1 above)
Comment (janvandermeulen) [reply pyproject.toml]: You are right, it was unnecessary clutter, I removed it.

File: tests/test_preprocess.py (hunk 3 above)
Comment (janvandermeulen) [reply tests/test_preprocess.py:14]: This seems like a good improvement for later.

File: tests/test_integration.py (hunk 4 above)
Comment (janvandermeulen) [reply tests/test_integration.py:14]: Yeah that would probably work better with the grading scheme as well.

**Comment by sr1998** [comment]
Seems good

**Review by sr1998 — APPROVED** [review]

**Comment by MichaelChan20** [comment]
Looks good, we should probably discuss which markers we are going to use for the tests

**Review by MichaelChan20 — APPROVED** [review]

**Pull request merged**